esp-hal/src/rmt.rs

```diff
@@ -622,47 +631,47 @@ where
 {
     /// Wait for the transaction to complete
     pub fn wait(mut self) -> Result<C, (Error, C)> {
-        loop {
-            if <C as TxChannelInternal>::is_error() {
-                return Err((Error::TransmissionError, self.channel));
-            }
-
-            if self.index >= self.data.len() {
-                break;
-            }
+        let memsize =
+            constants::RMT_CHANNEL_RAM_SIZE * <C as TxChannelInternal>::memsize() as usize;
 
+        while !self.remaining_data.is_empty() {
             // wait for TX-THR
-            while !<C as TxChannelInternal>::is_threshold_set() {}
+            while !<C as TxChannelInternal>::is_threshold_set() {
+                if <C as TxChannelInternal>::is_done() {
+                    // Unexpectedly done, even though we have data left: For example, this could
+                    // happen if there is a stop code inside the data and not just at the end.
+                    return Err((Error::TransmissionError, self.channel));
+                }
+                if <C as TxChannelInternal>::is_error() {
+                    // Not sure that this can happen? In any case, be sure that we don't lock up
+                    // here in case it can.
+                    return Err((Error::TransmissionError, self.channel));
+                }
+            }
             <C as TxChannelInternal>::reset_threshold_set();
 
             // re-fill TX RAM
-            let ram_index = (((self.index
-                - constants::RMT_CHANNEL_RAM_SIZE * <C as TxChannelInternal>::memsize() as usize)
-                / (constants::RMT_CHANNEL_RAM_SIZE
-                    * <C as TxChannelInternal>::memsize() as usize
-                    / 2))
-                % 2)
-                * (constants::RMT_CHANNEL_RAM_SIZE * <C as TxChannelInternal>::memsize() as usize
-                    / 2);
-
-            let ptr = (constants::RMT_RAM_START
-                + C::CHANNEL as usize * constants::RMT_CHANNEL_RAM_SIZE * 4
-                + ram_index * 4) as *mut u32;
-            for (idx, entry) in self.data[self.index..]
+            let ptr = unsafe { channel_ram_start(C::CHANNEL).add(self.ram_index) };
+            let count = self
+                .remaining_data
                 .iter()
-                .take(
-                    constants::RMT_CHANNEL_RAM_SIZE * <C as TxChannelInternal>::memsize() as usize
-                        / 2,
-                )
+                .take(memsize / 2)
                 .enumerate()
-            {
-                unsafe {
-                    ptr.add(idx).write_volatile(*entry);
-                }
-            }
-
-            self.index +=
-                constants::RMT_CHANNEL_RAM_SIZE * <C as TxChannelInternal>::memsize() as usize / 2;
+                .map(|(idx, entry)| unsafe { ptr.add(idx).write_volatile(*entry) })
+                .count();
```

Comment:
This makes sense but reads funny to me. Your intention is to avoid the need for an extra `let count = self.remaining_data.len().min(memsize / 2)` but I feel like the resulting iterator is a worse fit. For one, it's slightly surprising that you use `map` to actually write to the hardware. For the other, `count` *may* be less efficient if the compiler really turns it into `loop { write(); count += 1 }`.

Reply:
Yup, you got my intention right there; it also felt a little less error-prone due to not duplicating length calculation.
I agree regarding mutating in the iterator, that also feels unusual to me.

I'm pushing a modified version that is closer to the old code.